ledger/sandbox/src/main/scala/com/digitalasset/platform/store/ActiveLedgerState.scala

```diff
@@ -50,16 +50,16 @@ trait ActiveLedgerState[+Self] { this: ActiveLedgerState[Self] =>
   def lookupContractByKey(key: GlobalKey): Option[AbsoluteContractId]
 
   /** Called when a new contract is created */
-  def addContract(c: ActiveContract, keyO: Option[GlobalKey]): Self
+  def addContract(c: ActiveContract, keyO: Option[GlobalKey]): ALS
```

Comment:
Couldn't you just use `this.type` as the return values, and get rid of the generic-ness altogether?

Reply:
No. `this.type` is a singleton type and forces to return the very same instance on which the call was made, which is not necessarily the case in implementations (which may be immutable and thus need to return a copy to properly implement the `trait`, behavior-wise). With F-bounded polymorphism we're just restricting to the implementation type. It's more restrictive at compile time than the existing use of covariant return types and allows us to drop the usage of the implicit evidence.

Reply:
I'm convinced. 🙂